### PR TITLE
#3138 #3488 refactor ClansView

### PR DESCRIPTION
**Pull request by duybkict**

### CODE REVIEW

File: app/views/clans/ClansView.coffee [hunk 1]
```diff
@@ -52,12 +43,17 @@ module.exports = class ClansView extends RootView
...
     @supermodel.loadCollection(@publicClans, 'public_clans', {cache: false})
+    @publicClans = _.filter(@publicClans.models, (clan) -> clan.get('type') is 'public')
+
     @myClans = new CocoCollection([], { url: "/db/user/#{me.id}/clans", model: Clan, comparator: sortClanList })
     @listenTo @myClans, 'sync', =>
       @refreshNames @myClans.models
       @render?()
     @supermodel.loadCollection(@myClans, 'my_clans', {cache: false})
+    @myClans = @myClans.models
+
```
Comment (sderickson) [inline app/views/clans/ClansView.coffee]: This logic that was in `getRenderData` should go into a `onLoaded` function. Right now the logic runs after the `myClans` and `publicClans` collections have been created, but before their network requests have resolved. So they're empty at this point. Instead, put the logic in `onLoaded` like this:

```coffeescript
onLoaded: ->
  # prepare loaded data here
  super()
```

Any data manipulation happens after the data's been loaded, but before the page is re-rendered.

Also, view properties should not change in type, and setting `@myClans = @myClans.models` and `@publicClans = _.filter(@publicClans.models, ...)` changes those properties from collections to arrays. Either the view should use the collections (in which case `@publicClans` should be `reset` with the filtered models array, and the view should be updated to use collections) or the arrays should be set to new properties (`@publicClansArray` and `@myClansArray` for example).

Really appreciate you going through and standardizing views!

**Comment by duybkict** [comment]
I'm just started to getting to know CoffeScript and your code base, digging around like this would be the best approach for me. I will edit my code and make a new commit soon

**Comment by sderickson** [comment]
Sounds good!

There are many inconsistencies in the codebase, as we move generally toward using 3rd party tools the way they were intended, particularly in how we use Express on the server and Backbone on the client, and adding tests. When in doubt, [check the cookbook](https://github.com/codecombat/codecombat/wiki/Cookbook), which has up to date guidelines, or ask on feel free to ask on #general. The latter will help us figure out what to add to the Cookbook.

Thanks again!

File: app/views/clans/ClansView.coffee [hunk 2]
```diff
@@ -13,32 +13,30 @@ SubscribeModal = require 'views/core/SubscribeModal'
 module.exports = class ClansView extends RootView
   id: 'clans-view'
   template: template
+  publicClansArray: []
+  myClansArray: []
```
Comment (sderickson) [inline app/views/clans/ClansView.coffee]: As in `AnalyticsView` PR, one should never set mutable (and expected to be mutated) arrays or objects in class properties. They should be set in initialize to ensure each instance has their own.

**Comment by sderickson** [comment]
👍 
Squashed and merged in: 2ded5ff4f063a2dda4717b0cde6d65f1bbf1a432

**Pull request closed without merging**